{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Gene Read Matching: Parsing SAM/BAM and GTF Files\n",
        "This repository contains a bioinformatics workflow for parsing SAM/BAM files and comparing them with GTF gene annotation files to determine the number of reads aligned to specific genes. The workflow involves:\n",
        "\n",
        "Reading and parsing SAM/BAM files (containing aligned sequencing reads).\n",
        "Comparing the aligned reads to gene annotations (from GTF files), which are sorted by chromosome, exon start, and end positions.\n",
        "Counting how many reads align to each gene based on the gene's position.\n",
        "This repository uses example data (not actual experimental data) to demonstrate the workflow, with clear instructions on how to set up the pipeline. The focus is on methods for efficient parsing and comparison between large genomic datasets."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We first install gtfparse which is a Python library that efficiently parses and works with GTF (Gene Transfer Format) files and install pysam which is used to work with SAM (Sequence Alignment/Map), BAM (Binary Alignment/Map)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "98MK7bKllxHk",
        "outputId": "49fcd9d4-b22c-4a28-eff7-68366844b2aa"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting git+https://github.com/y9c/gtfparse.git\n",
            "  Cloning https://github.com/y9c/gtfparse.git to /tmp/pip-req-build-inl8wwdm\n",
            "  Running command git clone --filter=blob:none --quiet https://github.com/y9c/gtfparse.git /tmp/pip-req-build-inl8wwdm\n",
            "  Resolved https://github.com/y9c/gtfparse.git to commit 284a6b04aff70a69c82f939ed6da53322ad3854b\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: polars in /usr/local/lib/python3.10/dist-packages (from gtfparse==2.0.2) (0.17.3)\n",
            "Requirement already satisfied: typing_extensions>=4.0.1 in /usr/local/lib/python3.10/dist-packages (from polars->gtfparse==2.0.2) (4.5.0)\n",
            "Building wheels for collected packages: gtfparse\n",
            "  Building wheel for gtfparse (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for gtfparse: filename=gtfparse-2.0.2-py3-none-any.whl size=15190 sha256=1f4f337aa9cae53d85a95858b121cce86a320c71e9761c3b5166e1acbf6d68a5\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-689onrdh/wheels/70/aa/66/3693ca0d87954c0d972a3dc99feff3e14053dd9c97da24976f\n",
            "Successfully built gtfparse\n",
            "Installing collected packages: gtfparse\n",
            "Successfully installed gtfparse-2.0.2\n"
          ]
        }
      ],
      "source": [
        "!pip install git+https://github.com/y9c/gtfparse.git \n",
        "!pip install pysam #Installing pysam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We import the packages pandas, gtfparse and pysam so these tools can be later used in our analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "Wt-r7SFFmb6s"
      },
      "outputs": [],
      "source": [
        "import pandas as pd #Importing pandas\n",
        "from gtfparse import *\n",
        "from pysam import * #Importing everything from pysam"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We read the GTF file using the read_gtf command from the gtfparse package. Here, we label our sample data as sample.gtf and assign it to the variable df. Since we only need the exon data from the GTF file, we filter the original dataset and save the filtered data as df_genes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "zPXdQLthkz6Z"
      },
      "outputs": [],
      "source": [
        "#Reading the GTF file using the command from the gtfparse package\n",
        "df = read_gtf(\"sample.gtf\")\n",
        "#Filtering so we only get the exon data\n",
        "df_genes = df[df[\"feature\"]==\"exon\"]\n",
        "#Looking at the data\n",
        "df_genes"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We then filter the original DataFrame df to keep only exon entries from the first transcript (indicated by .1 in the transcript_id) and check if the filtering reduced the size of the dataset."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "t9GEhOmMEchn",
        "outputId": "ebcec454-344f-4bd2-9c27-b0908c0d36a8"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-0500091c-2c9b-4f4e-866f-65b10231e6e4\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>seqname</th>\n",
              "      <th>source</th>\n",
              "      <th>feature</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>score</th>\n",
              "      <th>strand</th>\n",
              "      <th>frame</th>\n",
              "      <th>geneID</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_name</th>\n",
              "      <th>transcript_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Chr1</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>3631</td>\n",
              "      <td>3913</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Chr1</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>3996</td>\n",
              "      <td>4276</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Chr1</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>4486</td>\n",
              "      <td>4605</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Chr1</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>4706</td>\n",
              "      <td>5095</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Chr1</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>5174</td>\n",
              "      <td>5326</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010</td>\n",
              "      <td>AT1G01010.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>414335</th>\n",
              "      <td>ChrM</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>360717</td>\n",
              "      <td>361052</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>ATMG01370</td>\n",
              "      <td>ATMG01370</td>\n",
              "      <td>ATMG01370.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>414337</th>\n",
              "      <td>ChrM</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>361062</td>\n",
              "      <td>361179</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>ATMG01380</td>\n",
              "      <td>ATMG01380</td>\n",
              "      <td>ATMG01380.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>414338</th>\n",
              "      <td>ChrM</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>361350</td>\n",
              "      <td>363284</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>ATMG01390</td>\n",
              "      <td>ATMG01390</td>\n",
              "      <td>ATMG01390.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>414339</th>\n",
              "      <td>ChrM</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>363725</td>\n",
              "      <td>364042</td>\n",
              "      <td>NaN</td>\n",
              "      <td>+</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>ATMG01400</td>\n",
              "      <td>ATMG01400</td>\n",
              "      <td>ATMG01400.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>414341</th>\n",
              "      <td>ChrM</td>\n",
              "      <td>TAIR10</td>\n",
              "      <td>exon</td>\n",
              "      <td>366086</td>\n",
              "      <td>366700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>None</td>\n",
              "      <td>ATMG01410</td>\n",
              "      <td>ATMG01410</td>\n",
              "      <td>ATMG01410.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>153730 rows Ã— 12 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0500091c-2c9b-4f4e-866f-65b10231e6e4')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-0500091c-2c9b-4f4e-866f-65b10231e6e4 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-0500091c-2c9b-4f4e-866f-65b10231e6e4');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-83ee30df-14cf-46f4-892e-4017d7047bcd\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-83ee30df-14cf-46f4-892e-4017d7047bcd')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-83ee30df-14cf-46f4-892e-4017d7047bcd button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "       seqname  source feature   start     end  score strand  frame geneID  \\\n",
              "0         Chr1  TAIR10    exon    3631    3913    NaN      +      0   None   \n",
              "1         Chr1  TAIR10    exon    3996    4276    NaN      +      0   None   \n",
              "2         Chr1  TAIR10    exon    4486    4605    NaN      +      0   None   \n",
              "3         Chr1  TAIR10    exon    4706    5095    NaN      +      0   None   \n",
              "4         Chr1  TAIR10    exon    5174    5326    NaN      +      0   None   \n",
              "...        ...     ...     ...     ...     ...    ...    ...    ...    ...   \n",
              "414335    ChrM  TAIR10    exon  360717  361052    NaN      -      0   None   \n",
              "414337    ChrM  TAIR10    exon  361062  361179    NaN      -      0   None   \n",
              "414338    ChrM  TAIR10    exon  361350  363284    NaN      -      0   None   \n",
              "414339    ChrM  TAIR10    exon  363725  364042    NaN      +      0   None   \n",
              "414341    ChrM  TAIR10    exon  366086  366700    NaN      -      0   None   \n",
              "\n",
              "          gene_id  gene_name transcript_id  \n",
              "0       AT1G01010  AT1G01010   AT1G01010.1  \n",
              "1       AT1G01010  AT1G01010   AT1G01010.1  \n",
              "2       AT1G01010  AT1G01010   AT1G01010.1  \n",
              "3       AT1G01010  AT1G01010   AT1G01010.1  \n",
              "4       AT1G01010  AT1G01010   AT1G01010.1  \n",
              "...           ...        ...           ...  \n",
              "414335  ATMG01370  ATMG01370   ATMG01370.1  \n",
              "414337  ATMG01380  ATMG01380   ATMG01380.1  \n",
              "414338  ATMG01390  ATMG01390   ATMG01390.1  \n",
              "414339  ATMG01400  ATMG01400   ATMG01400.1  \n",
              "414341  ATMG01410  ATMG01410   ATMG01410.1  \n",
              "\n",
              "[153730 rows x 12 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "#Using regular expression to filter so use only the exons that belong to the first transcript, their ID ends with a .1\n",
        "exons_data = df[(df[\"feature\"] == \"exon\") & df[\"transcript_id\"].str.match(r\".+\\.1$\")]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "We then open the sample BAM file, iterates over all reads, and extracts specific alignment information (chromosome, alignment start, and BAM file header) for each read, storing these details in reads_data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "7w-gaA9KChQ4"
      },
      "outputs": [],
      "source": [
        "#Opening BAM file using instructions from https://pysam.readthedocs.io/en/latest/api.html#sam-bam-cram-files\n",
        "bam_file = pysam.AlignmentFile(\"sample.bam\", \"rb\")\n",
        "#Iterating over the reads to get the chromosome name, and start of alignment and header from the bam file\n",
        "reads_data = [(read.reference_name, read.reference_start, read.header) for read in bam_file]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The code below goes through each exon in the exons_data DataFrame, counts the number of reads that align to the exon (using the BAM file), and stores the gene names and their corresponding read counts. If a chromosome is not found in the BAM file, the error is handled gracefully, and the script continues processing the next exon."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aAzWTuUGMq-7",
        "outputId": "017e947b-3006-4a27-bded-e95b0d789a01"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrC is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n",
            "ChrM is not found in the BAM file. Skip.\n"
          ]
        }
      ],
      "source": [
        "gene_names = [] #Creating a list of the genes\n",
        "read_counts = [] #Creating a list of the reads\n",
        "\n",
        "# Using a for loop to go through iterate through the rows and rename them\n",
        "for index, row in exons_data.iterrows():\n",
        "    exon_chrom = row[\"seqname\"]\n",
        "    exon_start = int(row[\"start\"])\n",
        "    exon_end = int(row[\"end\"])\n",
        "    gene_name = row[\"gene_name\"]\n",
        "\n",
        "    try: #We use try to avoid getting errors when chromosomes would not exist in the BAM file\n",
        "        # We count to count reads within the exon coordinates\n",
        "        #Souce: https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignmentFile.count\n",
        "        count = bam_file.count(exon_chrom, start=exon_start, stop=exon_end)\n",
        "\n",
        "\n",
        "        # Append gene name and read count to the list\n",
        "        gene_names.append(gene_name)\n",
        "        read_counts.append(count)\n",
        "\n",
        "\n",
        "    except ValueError as e:\n",
        "        # If we get a Key Error where the chromosome from the gtf does not exist in the BAM file, an error will print\n",
        "        #and the for loop will continue\n",
        "        print(f\"{exon_chrom} is not found in the BAM file. Skip.\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "The code first creates a DataFrame with gene names and the number of reads. It then groups by gene and sums the read counts for each gene. Finally, it prints the resulting DataFrame, which shows the total number of reads aligned to each gene."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GXv0XJQCT52y",
        "outputId": "b3548be6-aa9f-4984-aa63-3cd4256294f7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "            Gene  Number of Reads\n",
            "0      AT1G01010                3\n",
            "1      AT1G01020                1\n",
            "2      AT1G01030                0\n",
            "3      AT1G01040                9\n",
            "4      AT1G01046                0\n",
            "...          ...              ...\n",
            "33263  AT5G67600                0\n",
            "33264  AT5G67610                0\n",
            "33265  AT5G67620                0\n",
            "33266  AT5G67630                2\n",
            "33267  AT5G67640                1\n",
            "\n",
            "[33268 rows x 2 columns]\n"
          ]
        }
      ],
      "source": [
        "# Creating a DataFrame from the lists using pandas\n",
        "gene_counts_df = pd.DataFrame({\"Gene\": gene_names, \"Number of Reads\": read_counts})\n",
        "\n",
        "# Group by 'Gene' and sum the 'Number of Reads'\n",
        "gene_counts_combined = gene_counts_df.groupby('Gene')['Number of Reads'].sum().reset_index()\n",
        "\n",
        "# Print or use the resulting DataFrame\n",
        "print(gene_counts_combined)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
